pkg/daemon/controlhandler.go

```diff
@@ -0,0 +1,32 @@
+package daemon
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"connectrpc.com/connect"
+	controlv1 "github.com/codetent/crypta/gen/control/v1"
+	"github.com/codetent/crypta/gen/control/v1/controlv1connect"
+)
+
+type shutdowner func()
+
+type controlServiceServer struct {
+	controlv1connect.UnimplementedControlServiceHandler
+	shutdown shutdowner
+}
+
+func NewControlServiceHandler(shut shutdowner) (string, http.Handler) {
+	return controlv1connect.NewControlServiceHandler(&controlServiceServer{shutdown: shut})
+}
+
+func (s *controlServiceServer) Shutdown(
+	ctx context.Context,
+	req *connect.Request[controlv1.ShutdownRequest],
+) (*connect.Response[controlv1.ShutdownResponse], error) {
+	// wait so the response can be returned, and the server is shutdown afterwards
+	go time.AfterFunc(1*time.Second, s.shutdown)
```

Comment:
H'm it looks a bit errorprone because it can cause some race conditions.
This concept has some potential issues:
- Server shuts down and response is not sent (e.g. if message is not sent immediately)
- Response is sent but server does not shutdown because of any reason
- Client does not know either way if the shutdown was successful
Not sure if there is anything we can do other than passing waiting for the message to go out, and then passing the control to the shutdown function after closing the grpc server. Unfortunately also in this case the client has no ACK if it really happened.

What we could do instead:
- The client asks the server for its PID over grpc
- The client sends a signal like SIGTERM to the server
- The server catches the signal and shuts down gracefully
- The client checks the pid with a timeout to have an ACK

What do you think?

Reply:
Yes, good idea. I will try it out and update the PR.

Reply:
The problem I see with the approach is that it also contains raceconditions. For example:

Between 
- The server catches the signal and shuts down gracefully
- The client checks the pid with a timeout to have an ACK

The server might have already shutdown before the client checks the PID, and the PID might have been reused by the system for a new process. The client would therefore observe that the process is still alive, and has not been shutdown.

Reply:
Yes you are absolutely right. This could happen but I guess this can only happen if either it takes a lot of time in between or there are a lot of processes. I think it is pretty unlikely that it will directly reassign it.

Maybe we could inspect the process and store some information about it (e.g. command line, uptime or similar) before sending the signal. Then we could check again if it's the same process instead of just checking the pid.

There is also other potential race conditions between receiving the PID from the server, sending the kill signal and actually shutting down the server. If the endpoint would also return the other information (maybe as a hash to compare) we could also use it for identification later but in the end I guess you cannot completly get rid of all race conditions. We could just reduce the probability.